### DISCUSSION TITLE
Questions about `buffer_sv2` crate

### DISCUSSION THREAD

**Discussion by plebhash** [stratum-mining/stratum#1185] [General]
this discussion should guide the documentation of `buffer_sv2`, as described on https://github.com/stratum-mining/stratum/issues/845

**Comment by bit-aloo** [comment]
@Fi3, why is the shared_state in the buffer wrapped in an Arc around an atomic type? Atomics are already both Send and Sync compatible, and we're not accessing this value from multiple threads simultaneously. Is there a specific reason to have share reference?

**Reply by bit-aloo** [reply]
I see that the memory ordering is set to `Relaxed`, meaning there is no 'happens-before' relationship between atomic updates unless they occur on the same thread. If we intend to share a reference between different threads, Shouldn't the memory ordering for `shared_state` be stricter ? While it might default to acquire/release on x86_64, that’s not guaranteed on ARM, where more explicit ordering (such as Acquire/Release) is necessary to ensure correct behavior.

**Reply by Fi3** [reply]
I did it something like 2 years ago if not more, now I don't remember can you make an example where Relaxed mem ordering will broke the implementation? Like 
GoodState
threadA need to read because of ..  
...
BadState

thkns

**Reply by Fi3** [reply]
then whatever we end up with we will add a comment of why we chose that particular mem order
